### PR TITLE
[exporter/datadog] Simplify and add tags to logs from Datadog Exporter

### LINKED ISSUE #15387: [exporter/datadog] Add support for custom list of `attribute` -> `tag` conversion
### Is your feature request related to a problem? Please describe.

While it's great that the Datadog exporter is mapping from the [OpenTelemetry conventions](https://github.com/DataDog/datadog-agent/blob/main/pkg/otlp/model/attributes/attributes.go#L28), it seems that it is not possible to enrich telemetry with tags from the exporter on the collector.

### Describe the solution you'd like

An option key/value pairing on the exporter to define an `attribute` -> `tag` mapping. It could be something like:
```yaml
attributeTagMappings:
  - teamname: team
  - semver: version
```

### Describe alternatives you've considered

It seems like this mapping could also be achieved on [Logging Pipelines](https://docs.datadoghq.com/logs/log_configuration/pipelines). Would be preferable to keep all the configuration in the one exporter definition for us however. Likewise, it may be that there is agreement to use one specific tag, but each sender would source it from a different attribute, and a pipeline would need to be configured for all of those.

### Additional context

I'm not sure if these request should be on [Datadog Agent](https://github.com/DataDog/datadog-agent/), as that's where some changes will be required for this to be implement I guess.

### PR DESCRIPTION
**Pull request by liustanley**
**Description:**
- Adds additional mappings for trace ids and span ids
- Adds `otel_source:datadog_exporter` tag for logs sent from exporter
- Fixes #15387 by appending to tags in attributes instead of replacing them

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/15387

**Testing:** Added unit tests in `translator_test.go`

### CODE REVIEW

File: exporter/datadogexporter/internal/logs/translator.go [hunk 1]
```diff
@@ -83,6 +85,24 @@ func Transform(lr plog.LogRecord, res pcommon.Resource) datadogV2.HTTPLogItem {
 			l.Message = v.AsString()
 		case "status", "severity", "level", "syslog.severity":
 			status = v.AsString()
+		case "traceid", "contextmap.traceid", "oteltraceid":
+			ret, err := decodeTraceID(v.AsString())
+			if err != nil {
+				fmt.Errorf("failed to decode trace id %v: %v", v.AsString(), err)
```
Comment (dineshg13) [inline exporter/datadogexporter/internal/logs/translator.go:91]: use logger.

File: exporter/datadogexporter/internal/logs/translator.go [hunk 2]
```diff
@@ -83,6 +85,24 @@ func Transform(lr plog.LogRecord, res pcommon.Resource) datadogV2.HTTPLogItem {
...
+			ret, err := decodeSpanID(v.AsString())
+			if err != nil {
+				fmt.Errorf("failed to decode span id %v: %v", v.AsString(), err)
+			}
+			l.AdditionalProperties[ddSpanID] = strconv.FormatUint(spanIDToUint64(ret), 10)
+			l.AdditionalProperties[otelSpanID] = v.AsString()
+		case "ddtags":
+			var tags = append(attributes.TagsFromAttributes(res.Attributes()), v.AsString(), "otel:true")
+			tagStr := strings.Join(tags, ",")
+			l.AdditionalProperties[k] = tagStr
```
Comment (dineshg13) [inline exporter/datadogexporter/internal/logs/translator.go:105]: here we need to set `Ddtags`

File: exporter/datadogexporter/internal/logs/translator.go [hunk 3]
```diff
@@ -123,11 +143,9 @@ func Transform(lr plog.LogRecord, res pcommon.Resource) datadogV2.HTTPLogItem {
...
 	}
 
-	var tags = attributes.TagsFromAttributes(res.Attributes())
-	if len(tags) > 0 {
-		tagStr := strings.Join(tags, ",")
-		l.Ddtags = datadog.PtrString(tagStr)
-	}
+	var tags = append(attributes.TagsFromAttributes(res.Attributes()), "otel:true")
+	tagStr := strings.Join(tags, ",")
+	l.Ddtags = datadog.PtrString(tagStr)
```
Comment (dineshg13) [inline exporter/datadogexporter/internal/logs/translator.go:148]: we need to set this , only if it not already set

Comment (dineshg13) [reply exporter/datadogexporter/internal/logs/translator.go:148]: Interestingly the reason the test `append_tags` is passing because the `ddtags` from additional properties is used later in the method.  https://github.com/DataDog/datadog-api-client-go/blob/92a9c3eb71121aa596b15b3f1e8099d3c1c0a4c5/api/datadogV2/model_http_log_item.go#L208 . 
If the order changes in the marshal method , this would fail. we should explicitly set this only once.

File: exporter/datadogexporter/internal/logs/translator.go [hunk 4]
```diff
@@ -155,6 +173,18 @@ func extractHostNameAndServiceName(resourceAttrs pcommon.Map, logAttrs pcommon.M
 	return host, service
 }
 
+func decodeTraceID(traceID string) ([16]byte, error) {
+	var ret [16]byte
```
Comment (dineshg13) [inline exporter/datadogexporter/internal/logs/translator.go:177]: What would happen if the traceID is less than 16 bytes long ? can we add test for same.

File: exporter/datadogexporter/internal/logs/translator.go [hunk 5]
```diff
@@ -83,6 +85,24 @@ func Transform(lr plog.LogRecord, res pcommon.Resource) datadogV2.HTTPLogItem {
...
+			}
+			l.AdditionalProperties[ddSpanID] = strconv.FormatUint(spanIDToUint64(ret), 10)
+			l.AdditionalProperties[otelSpanID] = v.AsString()
+		case "ddtags":
+			var tags = append(attributes.TagsFromAttributes(res.Attributes()), v.AsString(), "otel:true")
+			tagStr := strings.Join(tags, ",")
+			l.AdditionalProperties[k] = tagStr
 		default:
 			l.AdditionalProperties[k] = v.AsString()
 		}
```
Comment (dineshg13) [inline exporter/datadogexporter/internal/logs/translator.go:108]: We need to check if ddTraceID is already set, and set it only if it is not set.

File: exporter/datadogexporter/internal/logs/translator.go [hunk 6]
```diff
@@ -83,6 +85,24 @@ func Transform(lr plog.LogRecord, res pcommon.Resource) datadogV2.HTTPLogItem {
...
+			l.AdditionalProperties[otelTraceID] = v.AsString()
+		case "spanid", "contextmap.spanid", "otelspanid":
+			ret, err := decodeSpanID(v.AsString())
+			if err != nil {
+				fmt.Errorf("failed to decode span id %v: %v", v.AsString(), err)
+			}
+			l.AdditionalProperties[ddSpanID] = strconv.FormatUint(spanIDToUint64(ret), 10)
+			l.AdditionalProperties[otelSpanID] = v.AsString()
+		case "ddtags":
+			var tags = append(attributes.TagsFromAttributes(res.Attributes()), v.AsString(), "otel:true")
```
Comment (dineshg13) [inline exporter/datadogexporter/internal/logs/translator.go:103]: Make this a constant. And a comment for same.

**Review by dineshg13 — CHANGES_REQUESTED** [review]

File: exporter/datadogexporter/internal/logs/translator.go (hunk 1 above)
Comment (dineshg13) [reply exporter/datadogexporter/internal/logs/translator.go:91]: and continue as well.

File: exporter/datadogexporter/internal/logs/translator.go [hunk 7]
```diff
@@ -58,9 +60,12 @@ const (
 	logLevelFatal = "fatal"
 )
 
+// This constant specifies a tag to be added to all logs sent from the Datadog exporter
+const otelTag = "otel:true"
```
Comment (mx-psi) [inline exporter/datadogexporter/internal/logs/translator.go:64]: nit: I would make this `source:otel` instead. I don't think we would ever have `otel:false` (rather, the absence of `otel:true` is what marks something as not coming from OpenTelemetry), but we could add other `source`s.

File: exporter/datadogexporter/internal/logs/translator.go [hunk 8]
```diff
@@ -83,6 +88,30 @@ func Transform(lr plog.LogRecord, res pcommon.Resource) datadogV2.HTTPLogItem {
 			l.Message = v.AsString()
 		case "status", "severity", "level", "syslog.severity":
 			status = v.AsString()
+		case "traceid", "contextmap.traceid", "oteltraceid":
+			ret, err := decodeTraceID(v.AsString())
+			if err != nil {
```
Comment (mx-psi) [inline exporter/datadogexporter/internal/logs/translator.go:93]: nit: declare the variables on the `if` statement to reduce their scope. Also, I would use something like `traceID` for readability, it's otherwise unclear to me what `ret` is
```suggestion
			if traceID, err := decodeTraceID(v.AsString()); err != nil {
```

File: exporter/datadogexporter/internal/logs/translator.go [hunk 9]
```diff
@@ -83,6 +88,30 @@ func Transform(lr plog.LogRecord, res pcommon.Resource) datadogV2.HTTPLogItem {
...
+				logger.Error("failed to decode trace id",
+					zap.String("trace_id", v.AsString()),
+					zap.Error(err))
+			} else if l.AdditionalProperties[ddTraceID] == "" {
+				l.AdditionalProperties[ddTraceID] = strconv.FormatUint(traceIDToUint64(ret), 10)
+				l.AdditionalProperties[otelTraceID] = v.AsString()
+			}
+		case "spanid", "contextmap.spanid", "otelspanid":
+			ret, err := decodeSpanID(v.AsString())
+			if err != nil {
```
Comment (mx-psi) [inline exporter/datadogexporter/internal/logs/translator.go:103]: ditto

File: exporter/datadogexporter/go.mod [hunk 10]
```diff
@@ -52,6 +52,7 @@ require (
 	github.com/Showmax/go-fqdn v1.0.0 // indirect
 	github.com/antonmedv/expr v1.9.0 // indirect
 	github.com/armon/go-metrics v0.3.10 // indirect
+	github.com/benbjohnson/clock v1.3.0 // indirect
```
Comment (dineshg13) [inline exporter/datadogexporter/go.mod:55]: Why is this being added ?

File: exporter/datadogexporter/internal/logs/translator.go [hunk 11]
```diff
@@ -58,9 +60,12 @@ const (
 	logLevelFatal = "fatal"
 )
 
+// This constant specifies a tag to be added to all logs sent from the Datadog exporter
```
Comment (dineshg13) [inline exporter/datadogexporter/internal/logs/translator.go:63]: comment should start with `otelTag`

File: exporter/datadogexporter/internal/logs/translator.go [hunk 12]
```diff
@@ -83,6 +88,28 @@ func Transform(lr plog.LogRecord, res pcommon.Resource) datadogV2.HTTPLogItem {
...
 			status = v.AsString()
+		case "traceid", "contextmap.traceid", "oteltraceid":
+			if traceID, err := decodeTraceID(v.AsString()); err != nil {
+				logger.Error("failed to decode trace id",
+					zap.String("trace_id", v.AsString()),
+					zap.Error(err))
+			} else if l.AdditionalProperties[ddTraceID] == "" {
+				l.AdditionalProperties[ddTraceID] = strconv.FormatUint(traceIDToUint64(traceID), 10)
+				l.AdditionalProperties[otelTraceID] = v.AsString()
+			}
```
Comment (dineshg13) [inline exporter/datadogexporter/internal/logs/translator.go:99]: no need of logging the error. set the properties when err == nil . Avoid if-else as much as possible.

File: exporter/datadogexporter/internal/logs/sender.go [hunk 13]
```diff
@@ -61,6 +61,9 @@ func NewSender(endpoint string, logger *zap.Logger, s exporterhelper.TimeoutSett
 // SubmitLogs submits the logs contained in payload to the Datadog intake
 func (s *Sender) SubmitLogs(ctx context.Context, payload []datadogV2.HTTPLogItem) error {
 	s.logger.Debug("Submitting logs", zap.Any("payload", payload))
+	if payload[0].HasDdtags() {
+		s.opts.Ddtags = datadog.PtrString(payload[0].GetDdtags())
+	}
```
Comment (dineshg13) [inline exporter/datadogexporter/internal/logs/sender.go:66]: Can you add comment why is this necessary ? Instead of replacing `s.opts.Ddtag` should we append to it ?

**Review by dineshg13 — CHANGES_REQUESTED** [review]

File: exporter/datadogexporter/go.mod (hunk 10 above)
Comment (liustanley) [reply exporter/datadogexporter/go.mod:55]: This got added when I imported `"go.uber.org/zap/zaptest"` in `translator_test.go`

**Review by dineshg13 — APPROVED** [review]
LTGM

**Review by mx-psi — APPROVED** [review]
